### DISCUSSION TITLE
ItemControl items separator

### DISCUSSION THREAD

**Discussion by SirZeeno** [AvaloniaUI/Avalonia#16528] [Q&A] [answered]
I was wondering what options I have to implement an items separator like a comma between items. I have looked to see what I could find on that topic and the one post I have seen about that was implementing Datatemplate.Triggers which avalonia doesn't implement not do I believe that WPF still implements it after 14 years. On the other hand I have tried using converters with that would return a bool depending on the index of the object in the Observablecollection, and if its the last then it returns false. All of it without success, Any suggestion is greatly appreciated :)

**Converter**
````
public class ListSeparatorConverter : IValueConverter
{
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        var currentItem = parameter;
        if (value is IList items && currentItem != null)
        {
            return items.IndexOf(currentItem) != items.Count - 1;
        }

        return false;
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        throw new NotImplementedException();
    }
}
````

**axaml**
````
<ItemsControl ItemsSource="{CompiledBinding Genres, Converter={StaticResource HashSetToObservableConverter}}" x:Name="Genres">
                                                <ItemsControl.ItemTemplate>
                                                    <DataTemplate DataType="animeArchiveHandlerGui:GenresDto">
                                                        <TextBlock Text="{CompiledBinding Name}"/>
                                                    </DataTemplate>
                                                </ItemsControl.ItemTemplate>
                                                <ItemsControl.ItemsPanel>
                                                    <ItemsPanelTemplate>
                                                        <StackPanel Orientation="Horizontal">
                                                            <TextBlock Text="{CompiledBinding Genres}" />
                                                            <TextBlock Text=", " IsVisible="{CompiledBinding Genres, Converter={StaticResource ListSeparatorConverter}, ConverterParameter={CompiledBinding Genres, Converter={StaticResource HashSetToObservableConverter}}}"/>
                                                        </StackPanel>
                                                    </ItemsPanelTemplate>
                                                </ItemsControl.ItemsPanel>
                                            </ItemsControl>
````

**Comment by IanRawley** [comment, marked as answer]
Unfortunately you can't bind to ConverterParameter, so you'll need to use a multibinding and IMultiValueConverter instead for the visibility of the ", " Textblock.  

I also have no idea if that ItemsPanelTemplate will work, I've never tried anything more complicated than swapping different panel types.  You might be better off either making your HashSetToObservableConverter insert dummy separator objects in the returned collection, although no idea how that would work, or having the TextBlock in the DataTemplate for GenresDto multibind to the GenreDto and the Genres list with a converter that returns a string of the name or name + ',' as appropriate.

**Reply by SirZeeno** [reply]
I solved the issue, I was just being an idiot and you are right, the ItemPanelTemplate was the issue and all I had to do was move what I had in the template into the ItemTemplate section. Small oversight on my end, because I was thinking that it would also work if I put them into the ItemsPanelTemplate.
